kem/frodo/frodo640shake/matrix_shake.go

```diff
@@ -0,0 +1,64 @@
+package frodo640shake
+
+import (
+	"github.com/cloudflare/circl/internal/sha3"
+)
+
+func expandSeedIntoA(A *[paramN * paramN]uint16, seed *[seedASize]byte, xof *sha3.State) error {
+	var ARow [paramN * 2]byte
+	var seedSeparated [2 + seedASize]byte
+
+	copy(seedSeparated[2:], seed[:])
+
+	for i := 0; i < paramN; i++ {
+		seedSeparated[0] = byte(i)
+		seedSeparated[1] = byte(i >> 8)
+
+		xof.Reset()
+		_, err := xof.Write(seedSeparated[:])
+		if err != nil {
+			return err
+		}
+		_, err = xof.Read(ARow[:])
+		if err != nil {
+			return err
+		}
+
+		for j := 0; j < paramN; j++ {
+			A[(i*paramN)+j] = uint16(ARow[j*2]) | (uint16(ARow[(j*2)+1]) << 8)
+		}
+	}
+	return nil
+}
+
+func mulAddASPlusE(out *[paramN * paramNbar]uint16, s []uint16, e []uint16, A *[paramN * paramN]uint16) {
```

Comment:
we can sort the parameters in order to differentiate the left-side vs the right-side multiplication
 mulAddASPlusE( out, A, S, E )
 mulAddSAPlusE( out, S, A, E )